### PR TITLE
[WIP]: Generalized Transfer Learning Across MolNet

### PR DESCRIPTION
**Pull request by vsomnath**
1. EarlyStoppingCallback + polished version of transfer-learning scripts
2. Need to add tests for EarlyStopping + add the download. 

Added an additional `stop_training` attribute to `KerasModel` which is updated by `EarlyStoppingCallback` & `compute_loss`, which computes the data-fit loss term.

### CODE REVIEW

File: deepchem/transfer_learning/chemnet.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+  ]
+
+  model = model_class(n_tasks=len(tasks), **model_hparams)
+  model.fit(train, nb_epoch, checkpoint_interval=0, callbacks=callbacks)
+  test_scores = model.evaluate(test, metrics=[metric], [])
+  logger.info("Test-{}: {}".format(metric.name, test_scores[metric.name]))
+  return model
+
+
+def finetune_model(dataset_name,
```
Comment (rbharath) [inline deepchem/transfer_learning/chemnet.py:89]: Could we generalize finetune to work on a user specified dataset? This would be a really useful features for users working with live small datasets

Comment (vsomnath) [reply deepchem/transfer_learning/chemnet.py:89]: Yes the arguments should be train, valid, test datasets then instead of `dataset_name` and `config`.

**Comment by peastman** [comment]
I don't see the value in EarlyStoppingCallback being separate from ValidationCallback.  You just took the ValidationCallback code, cloned it, and added more features while keeping all of its existing features.  So why have two classes at all?

Instead what about just adding an `early_stopping_metric` argument to ValidationCallback?  This can be either an int (the index of the metric to use) or 'loss'.  Or if it's None (the default), it doesn't do early stopping.

File: deepchem/models/keras_model.py [hunk 2]
```diff
@@ -147,6 +147,7 @@ def __init__(self,
       self.optimizer = optimizer
     self.tensorboard = tensorboard
     self.tensorboard_log_frequency = tensorboard_log_frequency
+    self.stop_training = False
```
Comment (peastman) [inline deepchem/models/keras_model.py:150]: Instead of using an internal field like this to signal the model to stop, how about having the callback raise a `StopIteration` exception?

File: deepchem/models/keras_model.py [hunk 3]
```diff
@@ -957,6 +961,33 @@ def get_global_step(self):
...
+    transformers: list of dc.trans.Transformers
+      Transformers that the dataset has been transformed by. The output
+      is passed through these transformers to undo the transformations.
+    """
+    if hasattr(self, 'mode'):
+      if self.mode == 'regression':
+        y_true = undo_transforms(dataset.y, transformers=transformers)
+      else:
+        y_true = to_one_hot(dataset.y.flatten(), self.n_classes).reshape(
+            -1, len(self.n_tasks), self.n_classes)
```
Comment (peastman) [inline deepchem/models/keras_model.py:980]: This is incorrect.  We shouldn't be hardcoding transforms like this.  The logic for computing the loss should exactly match what's done in `fit_generator()`.

Comment (vsomnath) [reply deepchem/models/keras_model.py:980]: I was confused about this: whether evaluate the loss on the transformed output, or the base one. I decided to stick the base output, but I will change this.

File: deepchem/models/keras_model.py (hunk 2 above)
Comment (vsomnath) [reply deepchem/models/keras_model.py:150]: would this look like:
```
for c in callbacks:
   try:
       c(model, step)
   except StopIteration:
       break
```

`stop_training=False` was also borrowed from Keras's design.

**Comment by vsomnath** [comment]
The main reason behind having separate validation and early stopping callbacks was to allow for different intervals for calling each of them. I agree though that a lot of code overlaps, and I will find a way to subclass. I also looked at Keras's design before doing this, and they have a ton of callbacks all in different classes (https://github.com/keras-team/keras/blob/master/keras/callbacks/callbacks.py).

File: deepchem/models/keras_model.py (hunk 2 above)
Comment (peastman) [reply deepchem/models/keras_model.py:150]: Exactly.

**Comment by peastman** [comment]
I don't think we even need to subclass.  The job of ValidationCallback is to evaluate metrics on the validation set.  One thing you can do with the results is to log them.  Another thing you can do is use the results for early stopping.  ValidationCallback can let you do both.

**Comment by vsomnath** [comment]
Apologies for the delay on getting to this! I have incorporated Peter's suggestions. If the basic layout looks okay, I will add tests and it should be ready for review.

File: deepchem/models/callbacks.py [hunk 4]
```diff
@@ -4,6 +4,13 @@
 
 import tensorflow as tf
 import sys
+from deepchem.metrics import Metric
+
+
+class StopIteration(Exception):
```
Comment (peastman) [inline deepchem/models/callbacks.py:10]: StopIteration is a [built in exception class](https://docs.python.org/3/library/exceptions.html#StopIteration).  We don't need to define our own class that shadows it.

File: deepchem/models/callbacks.py [hunk 5]
```diff
@@ -51,6 +61,14 @@ def __init__(self,
       if True, the best model is considered to be the one that minimizes the
       validation metric.  If False, the best model is considered to be the one
       that maximizes it.
+    early_stop_metric: str/int, default None
+      Indicates whether early stopping needs to be applied. If int, the index 
+      of the metric to use when checking for early stopping. If str, the loss
```
Comment (peastman) [inline deepchem/models/callbacks.py:66]: That should say, "If 'loss'" rather than, "If str".

File: deepchem/models/callbacks.py (hunk 4 above)
Comment (vsomnath) [reply deepchem/models/callbacks.py:10]: Oh good, I was not aware of this.

Comment (vsomnath) [reply deepchem/models/callbacks.py]: Done.

File: deepchem/models/callbacks.py [hunk 6]
```diff
@@ -51,6 +61,14 @@ def __init__(self,
...
       validation metric.  If False, the best model is considered to be the one
       that maximizes it.
+    early_stop_metric: str/int, default None
+      Indicates whether early stopping needs to be applied. If int, the index 
+      of the metric to use when checking for early stopping. If str, the loss 
+      on validation dataset is computed.
+    delta: float, default 0.01
+      Threshold for measuring the change in metric, to focus only on significant changes.
+    patience: int, default 0,
+      Number of steps with no improvement after which early stopping is enforced.
```
Comment (peastman) [inline deepchem/models/callbacks.py:65]: On both these options, how about adding the following sentence?  "If early_stop_metric is None, this is ignored."

The descriptions could also be a little clearer.  For `delta` maybe add the sentence, "Early stopping will occur when the metric improves by less than this amount."  And for `patience` say, "Number of successive steps..." to clarify that it's counting successive steps, not total steps.

File: deepchem/models/keras_model.py [hunk 7]
```diff
@@ -16,6 +16,7 @@
 from deepchem.models.optimizers import Adam
 from deepchem.trans import undo_transforms
 from deepchem.utils.evaluate import GeneratorEvaluator
+from deepchem.models.callbacks import StopIteration
```
Comment (peastman) [inline deepchem/models/keras_model.py]: Remove this line too so we use the built in exception type.

File: deepchem/models/keras_model.py [hunk 8]
```diff
@@ -957,6 +961,25 @@ def get_global_step(self):
...
+  def compute_loss(self, dataset):
+    """Computes the data-fit loss term on the given dataset.
+
+    Parameters
+    ----------
+    dataset: dc.data.Dataset,
+        Dataset to compute loss on
+    transformers: list of dc.trans.Transformers
+      Transformers that the dataset has been transformed by. The output
+      is passed through these transformers to undo the transformations.
```
Comment (peastman) [inline deepchem/models/keras_model.py:972]: I think this was left in by mistake.  There's not such argument.

File: deepchem/models/keras_model.py [hunk 9]
```diff
@@ -957,6 +961,25 @@ def get_global_step(self):
...
+    """
+    y_pred = self.predict(dataset)
+    y_pred = tf.convert_to_tensor(y_pred, dtype=tf.float32)
+    y_true = tf.convert_to_tensor(y_true, dtype=tf.float32)
```
Comment (peastman) [inline deepchem/models/keras_model.py:976]: You're referencing y_true before it's been defined.  Did you mean `dataset.y`?

**Comment by peastman** [comment]
I think the early stopping is looking pretty good.  I made some minor suggestions, and it still needs test cases.  Otherwise, I think it's in good shape.

For the transfer learning code, I wonder if we want to make it a bit more generic?  As currently written, it feels more like an example than a feature.  For example, it's hardcoded to pretrain on one specific dataset.  It might make sense to split out the early stopping feature, get that merged, then open a separate issue to discuss possible designs for more general transfer learning classes.

**Comment by vsomnath** [comment]
Closing this and splitting it into two PRs.
